vlasovsolver/cpu_trans_pencils.cpp

```diff
@@ -0,0 +1,1170 @@
+               SpatialCell *ncell = mpiGrid[nbrPair.first];
+               if (!ncell) continue;
+               int distanceInRefinedCells = -nbrPair.second[dimension];
+               if (distanceInRefinedCells == *it) {
+                  if (foundNeighborsM.find(nbrPair.first) != foundNeighborsM.end()) continue;
+                  foundNeighborsM.insert(nbrPair.first);
+                  if (!mpiGrid.is_local(nbrPair.first)) {
+                     ccell->SpatialCell::parameters[CellParams::AMR_TRANSLATE_COMM_X+dimension] = true;
+                     break;
+                  }
+               }
+            } // end loop over neighbors
+            iSrc--;
+         } // end loop over negative distances
+      } // end loop over local propagated cells
+   } // end loop over dimensions
+   return;
+}
+
+/* Get cellIDs for spatial cells that are considered target / source cells for a pencil.
+ *
+ * Source cells are cells that the pencil reads data from to compute polynomial
+ * fits that are used for propagation in the vlasov solver. All cells included
+ * in the pencil proper + VLASOV_STENCIL_WIDTH cells on both ends are source cells.
+ * Invalid cells are replaced by closest good cells.
+ * Boundary cells are included.
+ *
+ * Target cells are cells that the pencil writes data into after translation by
+ * the vlasov solver. All cells included in the pencil proper + 1 cells on both ends
+ * are target cells.
+ *
+ * There is only one list of cellIDs for the pencil, where each source cell has also
+ * their width stored, and for target cells, the relative contribution is stored. If
+ * the cell in question is not a target cell, the contribution is set to zero.
+ *
+ * @param [in] mpiGrid DCCRG grid object
+ * @param [inout] ids pointer to subsection of vector of ids for actual pencil
+ * @param [in] L length of pencil (including stencil cells)
+ * @param [in] dimension spatial dimension
+ * @param [in] path index of the desired face neighbor when going to a higher refinement level
+ * @param [out] source pointer to subsection of vector storing cell widths
+ * @param [out] targetRatios pointer to subsection of vector storing relative contribution of pencil to target cells
+ */
+void computeSpatialSourceCellsForPencil(const dccrg::Dccrg<SpatialCell,dccrg::Cartesian_Geometry>& mpiGrid,
+                                        CellID *ids,
+                                        const uint L,
+                                        const uint dimension,
+                                        std::vector<uint> path,
+                                        Realf* sourceDZ,
+                                        Realf* targetRatios
+                                        ){
+
+   // These neighborhoods now include the AMR addition beyond the regular vlasov stencil
+   int neighborhood = getNeighborhood(dimension,VLASOV_STENCIL_WIDTH);
+   stringstream ss;
+   for (uint j = 0; j < L; ++j) {
+      ss<< ids[j] << " ";
```

Comment:
Is it worth putting this behind some DEBUG flag? Maybe not as it's a rare thing anyway?

Reply:
It should never happen, is a fallthrough error.